### PR TITLE
Add support for functions to be parsed by the Lightning CLI in addition to Types

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Adds support for functions (e.g. classmethods) to be provided to the CLI in addition to types.
Needed for FlashCLI where we have data modules created from clasmethods

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/cli.py [hunk 1]
```diff
@@ -61,28 +61,40 @@ def __init__(self, *args: Any, parse_as_dict: bool = True, **kwargs: Any) -> Non
...
-            self.callback_keys.append(nested_key)
-        if subclass_mode:
-            return self.add_subclass_arguments(lightning_class, nested_key, required=True)
-        return self.add_class_arguments(
-            lightning_class,
-            nested_key,
-            fail_untyped=False,
-            instantiate=not issubclass(lightning_class, Trainer),
+        if isinstance(lightning_class,
+                      type) and issubclass(lightning_class, (Trainer, LightningModule, LightningDataModule, Callback)):
```
Comment (carmocca) [inline pytorch_lightning/utilities/cli.py:78]: I think using `inspect.isclass` would look more readable

File: pytorch_lightning/utilities/cli.py [hunk 2]
```diff
@@ -254,10 +269,10 @@ def __init__(
                 <https://jsonargparse.readthedocs.io/en/stable/#class-type-and-sub-classes>`_
                 of the given class.
         """
-        assert issubclass(trainer_class, Trainer)
-        assert issubclass(model_class, LightningModule)
+        assert issubclass(trainer_class, Trainer) if isinstance(trainer_class, type) else True
+        assert issubclass(model_class, LightningModule) if isinstance(model_class, type) else True
         if datamodule_class is not None:
-            assert issubclass(datamodule_class, LightningDataModule)
+            assert issubclass(datamodule_class, LightningDataModule) if isinstance(datamodule_class, type) else True
```
Comment (carmocca) [inline pytorch_lightning/utilities/cli.py:275]: I think i'd prefer removing these assertions since we already check in `add_lightning_class_args`

File: pytorch_lightning/utilities/cli.py [hunk 3]
```diff
@@ -333,8 +348,21 @@ def before_instantiate_classes(self) -> None:
...
         """Instantiates the classes using settings from self.config"""
         self.config_init = self.parser.instantiate_classes(self.config)
-        self.datamodule = self.config_init.get('data')
-        self.model = self.config_init['model']
+
+        datamodule_config = self.config_init.get('data')
+        if isinstance(self.datamodule_class, type):
+            self.datamodule = datamodule_config
+        elif self.datamodule_class is not None and datamodule_config is not None:
+            self.datamodule = self.datamodule_class(**datamodule_config)
```
Comment (carmocca) [inline pytorch_lightning/utilities/cli.py:356]: @mauvilsa could `add_function_arguments` have a flag similar to `instantiate` so we avoid this?

This doesn't block this PR, just as an off comment

File: pytorch_lightning/utilities/cli.py (hunk 2 above)
Comment (ethanwharris) [reply pytorch_lightning/utilities/cli.py]: Done :smiley:

File: pytorch_lightning/utilities/cli.py (hunk 1 above)
Comment (ethanwharris) [reply pytorch_lightning/utilities/cli.py]: Yeah, didn't know that was a thing haha, done!

File: pytorch_lightning/utilities/cli.py (hunk 3 above)
Comment (mauvilsa) [reply pytorch_lightning/utilities/cli.py:354]: I haven't looked in detail at all these changes. But I think as is now the argument linking would not work. Can this wait a bit for me to analyze it?

File: pytorch_lightning/utilities/cli.py [hunk 4]
```diff
@@ -344,7 +370,7 @@ def instantiate_trainer(self) -> None:
         callbacks = [self.config_init[c] for c in self.parser.callback_keys]
         self.config_init['trainer']['callbacks'].extend(callbacks)
         if 'callbacks' in self.trainer_defaults:
-            if isinstance(self.trainer_defaults['callbacks'], list):
+            if inspect.isclass(self.trainer_defaults['callbacks']):
```
Comment (mauvilsa) [inline pytorch_lightning/utilities/cli.py:373]: This is wrong. Checking that it is a list is not the same as checking if it is a class.

**Review by mauvilsa — CHANGES_REQUESTED** [review]

File: pytorch_lightning/utilities/cli.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/utilities/cli.py:354]: Sure, we aren't targeting 1.4 for this

File: pytorch_lightning/utilities/cli.py (hunk 4 above)
Comment (ethanwharris) [reply pytorch_lightning/utilities/cli.py]: Whoops just an accidental bug, I didn't think those two checks were equivalent haha. Fixed now

**Comment by carmocca** [comment]
There are some mypy complains to adress

**Comment by ethanwharris** [comment]
@carmocca Yeah, mypy wasn't picking up `inspect.isclass` for some reason. Fixed now :smiley:

File: pytorch_lightning/utilities/cli.py (hunk 3 above)
Comment (mauvilsa) [reply pytorch_lightning/utilities/cli.py:354]: I propose the following. jsonargparse would have a new function `class_from_function`. With this instead of having two cases, only `add_class_arguments` would be used and not `add_function_arguments`. At the start of `add_lightning_class_args` it would be like:

```python
if callable(lightning_class):
    lightning_class = class_from_function(lightning_class)
```

The callable would be expected to have a return type annotation such that the `issubclass` after prevents giving a function that returns an unexpected class. Since only `add_class_arguments` would be used, then `self.datamodule_class(**datamodule_config)` is not necessary and the linking of arguments would work correctly.

Comment (carmocca) [reply pytorch_lightning/utilities/cli.py:354]: Sounds good to me

Comment (mauvilsa) [reply pytorch_lightning/utilities/cli.py:354]: `class_from_function` is now implemented and part of jsonargparse v3.17.0.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by mauvilsa — CHANGES_REQUESTED** [review]

**Comment by ethanwharris** [comment]
@mauvilsa thanks for the updates :smiley: I've updated the PR with the required changes, mind taking another look?

File: pytorch_lightning/utilities/cli.py [hunk 5]
```diff
@@ -61,28 +63,38 @@ def __init__(self, *args: Any, parse_as_dict: bool = True, **kwargs: Any) -> Non
...
-            nested_key,
-            fail_untyped=False,
-            instantiate=not issubclass(lightning_class, Trainer),
+        if callable(lightning_class) and not inspect.isclass(lightning_class):
+            lightning_class = class_from_function(lightning_class)
+
+        if inspect.isclass(lightning_class) and issubclass(
+            cast(type, lightning_class), (Trainer, LightningModule, LightningDataModule, Callback)
+        ):
+            if issubclass(cast(type, lightning_class), Callback):
```
Comment (carmocca) [inline pytorch_lightning/utilities/cli.py:85]: You only need to `cast(type, lightning_class)` once

**Review by carmocca — APPROVED** [review]

**Review by mauvilsa — APPROVED** [review]

File: pytorch_lightning/utilities/cli.py (hunk 5 above)
Comment (ethanwharris) [reply pytorch_lightning/utilities/cli.py]: Done :smiley:

File: pytorch_lightning/utilities/cli.py [hunk 6]
```diff
@@ -61,28 +63,39 @@ def __init__(self, *args: Any, parse_as_dict: bool = True, **kwargs: Any) -> Non
 
     def add_lightning_class_args(
         self,
-        lightning_class: Union[Type[Trainer], Type[LightningModule], Type[LightningDataModule], Type[Callback]],
+        lightning_class: Union[Callable[..., Union[Trainer, LightningModule, LightningDataModule, Callback]],
```
Comment (Borda) [inline pytorch_lightning/utilities/cli.py:66]: this typing is quite wild :D

File: tests/utilities/test_cli.py [hunk 7]
```diff
@@ -220,45 +220,54 @@ def test_init_from_argparse_args(cli_args, extra_args):
...
+
+def trainer_builder(
+    limit_train_batches: int,
+    fast_dev_run: bool = False,
+    callbacks: Optional[Union[List[Callback], Callback]] = None
+) -> Trainer:
+    return Trainer(limit_train_batches=limit_train_batches, fast_dev_run=fast_dev_run, callbacks=callbacks)
+
+
+@pytest.mark.parametrize(['trainer_class', 'model_class'], [(Trainer, Model), (trainer_builder, model_builder)])
```
Comment (Borda) [inline tests/utilities/test_cli.py:242]:
```suggestion
def _model_builder(model_param: int) -> Model:
    return Model(model_param)


def _trainer_builder(
    limit_train_batches: int,
    fast_dev_run: bool = False,
    callbacks: Optional[Union[List[Callback], Callback]] = None
) -> Trainer:
    return Trainer(limit_train_batches=limit_train_batches, fast_dev_run=fast_dev_run, callbacks=callbacks)


@pytest.mark.parametrize(['trainer_class', 'model_class'], [(Trainer, Model), (_trainer_builder, _model_builder)])
```

**Review by Borda — APPROVED** [review]
